pyscf/lib/vhf/nr_sgx_direct.c

```diff
@@ -35,138 +36,223 @@ typedef struct {
+                double* all_grids, int tot_grids)
 {
         DECLARE_ALL;
 
-        shls[2] = ksh0 + ksh;
-
-        for (ish = ish0; ish < ish1; ish++) {
-        for (jsh = jsh0; jsh <= ish; jsh++) {
-                shls[0] = ish;
-                shls[1] = jsh;
-                if ((*fprescreen)(shls, vhfopt, atm, bas, env)
-                    && (*intor)(buf, NULL, shls, atm, natm, bas, nbas, env,
-                                cintopt, cache)) {
-                        i0 = ao_loc[ish  ] - ioff;
-                        j0 = ao_loc[jsh  ] - joff;
-                        i1 = ao_loc[ish+1] - ioff;
-                        j1 = ao_loc[jsh+1] - joff;
-                        for (idm = 0; idm < n_dm; idm++) {
-                                jkop[idm]->contract(buf, dms[idm], vjk[idm],
-                                                    i0, i1, j0, j1, ksh);
+        i0 = ao_loc[ish  ] - ioff;
+        j0 = ao_loc[jsh  ] - joff;
+        i1 = ao_loc[ish+1] - ioff;
+        j1 = ao_loc[jsh+1] - joff;
+
+        int tmp_ngrids = 0;
+        int k;
+        int* inds = (int*) malloc(tot_grids*sizeof(int));
+
+        double *grids = env + (size_t) env[PTR_GRIDS];
+        
+        if (vhfopt != NULL && vhfopt->pscreen == 1) {
+                for (k = 0; k < tot_grids; k++) {
+                        shls[2] = k;
+                        if (SGXnr_pj_prescreen(shls, vhfopt, atm, bas, env)) {
+                                grids[3*tmp_ngrids+0] = all_grids[3*k+0];
+                                grids[3*tmp_ngrids+1] = all_grids[3*k+1];
+                                grids[3*tmp_ngrids+2] = all_grids[3*k+2];
+                                inds[tmp_ngrids] = k;
+                                tmp_ngrids++;
                         }
                 }
-        } }
+                env[NGRIDS] = tmp_ngrids;
+                /*for (k = 0; k < tot_grids; k++) {
+                        shls[2] = k;
+                        grids[3*tmp_ngrids+0] = all_grids[3*k+0];
+                        grids[3*tmp_ngrids+1] = all_grids[3*k+1];
+                        grids[3*tmp_ngrids+2] = all_grids[3*k+2];
+                        inds[tmp_ngrids] = k;
+                        tmp_ngrids++;
+                }
+                env[NGRIDS] = tmp_ngrids;*/
+        } else {
+                for (k = 0; k < tot_grids; k++) {
+                        shls[2] = k;
+                        grids[3*tmp_ngrids+0] = all_grids[3*k+0];
+                        grids[3*tmp_ngrids+1] = all_grids[3*k+1];
+                        grids[3*tmp_ngrids+2] = all_grids[3*k+2];
+                        inds[tmp_ngrids] = k;
+                        tmp_ngrids++;
+                }
+                env[NGRIDS] = tmp_ngrids;
+        }
+
+        int grid0, grid1;
+        const int dims[] = {ao_loc[ish+1]-ao_loc[ish], ao_loc[jsh+1]-ao_loc[jsh], tmp_ngrids};
+        for (grid0 = 0; grid0 < env[NGRIDS]; grid0 += BLKSIZE) {
```

Comment:
env[NGRIDS] has wrong type. use tmp_ngrids instead of env[NGRIDS]

Reply:
Note, I used ngrids here, not tmp_ngrids, which is only defined in a different scope.